### PR TITLE
Highlight JSX in TypeScript.

### PR DESCRIPTION
**Pull request by sdegutis**
This PR adds JSX highlighting to TypeScript natively using Monaco's Monarch tokenizer, along with a few other nice to have changes:

1. Highlight `...` as a keyword
2. Highlight `export/etc` as control-flow keywords
3. Highlight class fields
4. Highlight function/class names
5. Highlight var/let/const names
6. Highlight new Foo
7. Highlight function/method calls
8. Highlight JSX

It has been thoroughly tested on https://vanillajsx.com/monarch/ but it would be useful to have more people test for edge cases.

A few unrelated tests are failing such as PHP and Less, which I don't understand how they're related since my code is isolated to one file. The failures look to just be whitespace errors, but it's hard to tell.

I didn't update the TypeScript tests yet, so they're failing too.

### CODE REVIEW

File: src/basic-languages/typescript/typescript.ts [hunk 1]
```diff
@@ -222,143 +131,249 @@ export const language = {
...
+
+	  jsxIdent: [
+		[/\./, { token: 'delimiter', switchTo: '$S0^' }],
+		[/[A-Z][\w$]*/, { token: 'type.identifier', switchTo: '$S0$0' }],
+		[/[\w$-]+/, { token: 'tag', switchTo: '$S0$0' }],
+		[/.+/, { token: '@rematch', switchTo: '@$S2.$S3.$S4' }],
+	  ],
+
+	  jsxOpen: [
+		[/{/, { token: 'keyword', next: '@root.INJSX', bracket: '@open' }],
```
Comment (shihui-huang) [inline src/basic-languages/typescript/typescript.ts:264]: Hi @thesoftwarephilosopher, Thank you so much for sharing this token provider file. It's very helpful and straightforward, and it solved my problem very easily!

I noticed a small inconsistency where the first and second brackets had different tokens in some cases. I made a slight modification to make them consistent.
**Before:**
image
image
**After:**
image
image


```suggestion
		[/{/, { token: 'delimiter.curly', next: '@root.INJSX', bracket: '@open' }],
```

File: src/basic-languages/typescript/typescript.ts [hunk 2]
```diff
@@ -222,143 +131,249 @@ export const language = {
...
+		[/>/, { token: 'delimiter.html', switchTo: '@jsxText.$S2' }],
+		[/\/>/, { token: 'delimiter.html', next: '@pop' }],
+		[/ +([\w-$]+)/, 'attribute.name'],
+		[/(=)(')/, ['delimiter', { token: 'string', next: '@string_single' }]],
+		[/(=)(")/, ['delimiter', { token: 'string', next: '@string_double' }]],
+		[/(=)({)/, ['delimiter', { token: '@brackets', next: '@root.INJSX' }]],
+	  ],
+
+	  jsxText: [
+		[/{/, { token: 'keyword', next: '@root.INJSX', bracket: '@open' }],
```
Comment (shihui-huang) [inline src/basic-languages/typescript/typescript.ts:274]:
```suggestion
		[/{/, { token: 'delimiter.curly', next: '@root.INJSX', bracket: '@open' }],
```

File: src/basic-languages/typescript/typescript.ts [hunk 3]
```diff
@@ -222,143 +131,249 @@ export const language = {
...
+		  }
 		],
 
-		string_backtick: [
-			[/\$\{/, { token: 'delimiter.bracket', next: '@bracketCounting' }],
-			[/[^\\`$]+/, 'string'],
-			[/@escapes/, 'string.escape'],
-			[/\\./, 'string.escape.invalid'],
-			[/`/, 'string', '@pop']
+		[/\.\.\./, 'keyword'],
```
Comment (shihui-huang) [inline src/basic-languages/typescript/typescript.ts:222]: And I think it would be more fitting to categorize `...` as an operator

**Comment by raflymln** [comment]
Hey, is there any followup to this PR? @thesoftwarephilosopher

**Comment by thesoftwarephilosopher** [comment]
@microsoft-github-policy-service agree

**Comment by brianzinn** [comment]
community PR approvals `Started 101007m ago`!  Anyway this could be merged? It would be so useful to have built-in JSX/TSX highlighting. 😄
